### PR TITLE
[GPflow 2.0] Natural Gradient

### PR DESCRIPTION
**Pull request by awav**
Implementation of the natural gradient optimizer in GPflow 2.0 with tests and updated notebook.

### CODE REVIEW

File: tests/test_natural_gradient.py [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+    svgp.q_mu.trainable = True
+    svgp.q_sqrt.trainable = True
+    return sgpr, svgp
+
+
+def assert_gpr_vs_vgp(m1: tf.Module,
+                      m2: tf.Module,
+                      gamma: float = 1.0,
+                      maxiter: int = 1,
+                      xi_transform: gpflow.optimizers.natgrad.XiTransform = None):
```
Comment (stefanosele) [inline tests/test_natural_gradient.py]: shouldn't that be `Optional[gpflow.optimizers.natgrad.XiTransform]`?

File: gpflow/optimizers/natgrad.py [hunk 2]
```diff
@@ -0,0 +1,324 @@
...
+            booktitle={AISTATS},
+            year={2018}
+
+        In addition, for convenience with the rest of GPflow, this code computes ∂L/∂η using
+        the chain rule:
+
+        ∂L/∂η = (∂[q_μ, q_sqrt] / ∂η)(∂L / ∂[q_μ, q_sqrt])
+
+        In total there are three derivative calculations:
+        natgrad L w.r.t ξ  = (∂ξ / ∂nat) [ (∂[q_μ, q_sqrt] / ∂η)(∂L / ∂[q_μ, q_sqrt]) ]^T
```
Comment (awav) [inline gpflow/optimizers/natgrad.py]:
```suggestion
        natgrad L w.r.t ξ  = (∂ξ / ∂θ) [ (∂[q_μ, q_sqrt] / ∂η)(∂L / ∂[q_μ, q_sqrt]) ]^T
```

File: gpflow/optimizers/natgrad.py [hunk 3]
```diff
@@ -0,0 +1,324 @@
...
+                ```
+            :param feed_dict: Feed dictionary of tensors passed to session run method.
+            :param maxiter: Number of run interation. Default value: 1000.
+            :param anchor: Synchronize updated parameters for a session with internal
+                parameter's values.
+            :param step_callback: A callback function to execute at each optimization step.
+                The callback should accept variable argument list, where first argument is
+                optimization step number.
+            :type step_callback: Callable[[], None]
+            :param kwargs: Extra parameters passed to session run's method.
```
[the remark this answers is not in the document]
Comment (awav) [reply gpflow/optimizers/natgrad.py]: Nope, I will remove it.

File: gpflow/optimizers/natgrad.py [hunk 4]
```diff
@@ -0,0 +1,324 @@
...
+        Minimizes objective function of the model.
+        Natural Gradient optimizer works with variational parameters only.
+        There are two supported ways of transformation for parameters:
+            - XiNat
+            - XiSqrtMeanVar
+        Custom transformations are also possible, they should implement
+        `XiTransform` interface.
+
+            :param model: GPflow model.
+            :param session: Tensorflow session where optimization will be run.
```
[the remark this answers is not in the document]
Comment (awav) [reply gpflow/optimizers/natgrad.py]: I need to update docstrings

File: gpflow/optimizers/natgrad.py [hunk 5]
```diff
@@ -0,0 +1,324 @@
...
+                nat1, nat2 = meanvarsqrt_to_natural(q_mu, q_sqrt)
+                xi1_nat, xi2_nat = xi_transform.naturals_to_xi(nat1, nat2)
+                dummy_tensors = tf.ones_like(xi1_nat), tf.ones_like(xi2_nat)
+                with tf.GradientTape(watch_accessed_variables=False) as forward_tape:
+                    forward_tape.watch(dummy_tensors)
+                    dummy_gradients = tape.gradient([xi1_nat, xi2_nat], [nat1, nat2], output_gradients=dummy_tensors)
+
+        # 1) the oridinary gpflow gradient
+        dL_dmean, dL_dvarsqrt = tape.gradient(loss, [q_mu, q_sqrt])
+        dL_dvarsqrt = q_sqrt.transform.forward(dL_dvarsqrt)
```
[the remark this answers is not in the document]
Comment (awav) [reply gpflow/optimizers/natgrad.py]: We can only compute the gradient w.r.t unconstrained variable and we are lucky that transformation is just a reshuffling of the elements. In gpflow 1.0 we could take a gradient w.r.t. constrained tensor and therefore we didn't have to use forward transform.

File: gpflow/optimizers/natgrad.py [hunk 6]
```diff
@@ -0,0 +1,324 @@
...
+
+        xi1, xi2 = xi_transform.meanvarsqrt_to_xi(q_mu, q_sqrt)
+        xi1_new = xi1 - self.gamma * nat_dL_xi1
+        xi2_new = xi2 - self.gamma * nat_dL_xi2
+
+        # Transform back to the model parameters [q_μ, q_sqrt]
+        mean_new, varsqrt_new = xi_transform.xi_to_meanvarsqrt(xi1_new, xi2_new)
+
+        q_mu.assign(mean_new)
+        q_sqrt.assign(varsqrt_new)
```
[the remark this answers is not in the document]
Comment (awav) [reply gpflow/optimizers/natgrad.py]: The `assign` does necessary transformation (inverse of the bijector).

File: tests/test_natural_gradient.py (hunk 1 above)
Comment (awav) [reply tests/test_natural_gradient.py]: Yes, that should be the optional type.

**Pull request merged**